### ISSUE TITLE
(1-\beta) multiplying on learning rate instead of gradients while applying momentum in TF version QHM

### ISSUE THREAD

**Issue by allenwoods** [facebookresearch/qhoptim#2]
According to the paper, the moment should be
![image](https://user-images.githubusercontent.com/7327215/98514337-4145b380-22a4-11eb-8697-deba9f86736f.png)

and also in torch version
https://github.com/facebookresearch/qhoptim/blob/e81dea3f2765780cf4fbb90b87b22ba7604b8625/qhoptim/pyt/qhm.py#L133

However, the TF version has a different implement that set the learning rate to
https://github.com/facebookresearch/qhoptim/blob/e81dea3f2765780cf4fbb90b87b22ba7604b8625/qhoptim/tf/qhm.py#L131

Doesn't that makes the formula into
![image](https://user-images.githubusercontent.com/7327215/98514276-27a46c00-22a4-11eb-9b7d-f0d0550dde31.png)
?

Is this a typo, or am I misunderstand something?  TIA 😁

**Comment by jma127** [comment]
Hi @allenwoods , it's been a while since I last looked at the TF code but IIRC the tensorflow momentum buffer is updated as follows:

    g_{+ 1} = g + \nabla L

and thus we need to scale it by 1 - beta in order to ensure consistency with the paper.

**Comment by allenwoods** [comment]
Thanks for the quick reply！The TensorFlow momentum buffer is updated as follows：
```python
accum = accum * momentum + grad
var -= lr * accum
```
where `var` is the weights.

In that case, if the lr is `nu * (1.0 - momentum) * learning_rate`, it seems both momentum and grad are scaled by 1-beta, while the paper indicates we only need to scale the gradient. Therefore, I modified the momentum_op to
```python
        momentum_op = training_ops.resource_apply_momentum(
            var.handle,
            momentum_buffer.handle,
            nu * learning_rate,
            (1.0 - momentum) * grad,
            momentum,
            use_locking=self._use_locking,
            use_nesterov=False,
        )
```

Is it right?

**Comment by jma127** [comment]
I believe the code is correct as written. The reason you need to scale by the extra (1 - momentum) term is that the accum term as you describe it is "unnormalized" relative to the momentum buffer used by the paper.

**Comment by allenwoods** [comment]
I am confused, does that mean `momentum_buffer` in TF is 1/(1-momentum) of `momentum_buffer` in torch? Therefore, we need to apply (1 - momentum) term on both momentum and grad when using the TF version, while in the torch version, we only need to apply it to grad?

However, I couldn't find the code that leads such a difference in the torch version, since the only place multiplying  (1 - momentum) (as far as I can find) is https://github.com/facebookresearch/qhoptim/blob/e81dea3f2765780cf4fbb90b87b22ba7604b8625/qhoptim/tf/qhm.py#L131

**Comment by jma127** [comment]
See https://github.com/facebookresearch/qhoptim/blob/master/qhoptim/pyt/qhm.py#L133 .

**Comment by allenwoods** [comment]
That is exactly the line I reference in the first post, in which the momentum was multiplied **by `beta` only**, then added by `(1-beta)*grad`. Then the momentum was multiplied by `nu`.

However, in the [TF version code](https://github.com/facebookresearch/qhoptim/blob/e81dea3f2765780cf4fbb90b87b22ba7604b8625/qhoptim/tf/qhm.py#L128-L136)
```python
        momentum_op = training_ops.resource_apply_momentum(
            var.handle,
            momentum_buffer.handle,
            nu * (1.0 - momentum) * learning_rate,
            grad,
            momentum,
            use_locking=self._use_locking,
            use_nesterov=False,
        )
```
where **both** momentum and grad are multiplied by (1-beta)*nu. Therefore, the TF version's momentum is multiplied by `(1-beta)*beta` instead of `beta` before multiplied by `nu`.

Did I miss something?

**Comment by jma127** [comment]
Yes, you should read the specification of training_ops.resource_apply_momentum to understand the precise mathematical form.

I advise you to set up a toy problem in TF and verify the step-by-step operation of the algorithm by hand—I anticipate you will find that the parameters follow the form from the paper. If you find otherwise, please circle back and we can discuss.
